### ISSUE TITLE
Wrong result (1 byte too short) produced by Exiv2::base64decode

### ISSUE THREAD

**Issue by agriggio** [Exiv2/exiv2#1358]
Hi,
it seems that Exiv2::base64decode produces a wrong result, i.e. the output result is 1 byte too short. Attached a small example to reproduce, where I compare with the public domain implementation of base64decode taken from wikipedia (https://en.wikibooks.org/wiki/Algorithm_Implementation/Miscellaneous/Base64#C++_2).
(I've also tested with other implementations)

I'm using exiv2 0.27.3 on Linux

Sample code:

```c++
#include <iostream>
#include <vector>
#include <string>
#include <exiv2/exiv2.hpp>


// taken from https://en.wikibooks.org/wiki/Algorithm_Implementation/Miscellaneous/Base64#C++_2, license: public domain
#define TCHAR char
#define TEXT(x) x
#define DWORD long
#define BYTE unsigned char
 
const static TCHAR padCharacter = TEXT('=');
std::vector<BYTE> base64decode(const std::basic_string<TCHAR>& input)
{
    if (input.length() % 4) //Sanity check
        throw std::runtime_error("Non-Valid base64!");
    size_t padding = 0;
    if (input.length())
    {
        if (input[input.length()-1] == padCharacter)
            padding++;
        if (input[input.length()-2] == padCharacter)
            padding++;
    }
    //Setup a vector to hold the result
    std::vector<BYTE> decodedBytes;
    decodedBytes.reserve(((input.length()/4)*3) - padding);
    DWORD temp=0; //Holds decoded quanta
    std::basic_string<TCHAR>::const_iterator cursor = input.begin();
    while (cursor < input.end())
    {
        for (size_t quantumPosition = 0; quantumPosition < 4; quantumPosition++)
        {
            temp <<= 6;
            if       (*cursor >= 0x41 && *cursor <= 0x5A) // This area will need tweaking if
                temp |= *cursor - 0x41;		              // you are using an alternate alphabet
            else if  (*cursor >= 0x61 && *cursor <= 0x7A)
                temp |= *cursor - 0x47;
            else if  (*cursor >= 0x30 && *cursor <= 0x39)
                temp |= *cursor + 0x04;
            else if  (*cursor == 0x2B)
                temp |= 0x3E; //change to 0x2D for URL alphabet
            else if  (*cursor == 0x2F)
                temp |= 0x3F; //change to 0x5F for URL alphabet
            else if  (*cursor == padCharacter) //pad
            {
                switch( input.end() - cursor )
                {
                case 1: //One pad character
                    decodedBytes.push_back((temp >> 16) & 0x000000FF);
                    decodedBytes.push_back((temp >> 8 ) & 0x000000FF);
                    return decodedBytes;
                case 2: //Two pad characters
                    decodedBytes.push_back((temp >> 10) & 0x000000FF);
                    return decodedBytes;
                default:
                    throw std::runtime_error("Invalid Padding in Base 64!");
                }
            }  else
                throw std::runtime_error("Non-Valid Character in Base 64!");
            cursor++;
        }
        decodedBytes.push_back((temp >> 16) & 0x000000FF);
        decodedBytes.push_back((temp >> 8 ) & 0x000000FF);
        decodedBytes.push_back((temp      ) & 0x000000FF);
    }
    return decodedBytes;
}


int main()
{
    std::cout << "Exiv2 version is: " << Exiv2::version() << std::endl;
    
    std::string data = "yccA0gAAb9HiTEBWn1DOTEBW";
    std::vector<char> buf(data.size());
    long n = Exiv2::base64decode(data.c_str(), &(buf[0]), buf.size());
    if (n < 0) {
        std::cout << "ERROR!" << std::endl;
    }
    buf.resize(n);
    std::cout << buf.size() << ": ";
    for (auto c : buf) {
        std::cout << int(static_cast<unsigned char>(c)) << " ";
    }
    std::cout << std::endl;
    auto res = base64decode(data);
    std::cout << res.size() << ": ";
    for (auto c : res) {
        std::cout << int(c) << " ";
    }
    std::cout << std::endl;
    return 0;
}
```

**Comment by clanmills** [comment]
Thanks for reporting this.  I will investigate.

**Comment by clanmills** [comment]
I've looked at this and submitted a PR.  #1360.  I've asked another contributor to make changes to the test suite concerning this.  To verify the fix today, I cut'n'pasted your code into samples/exifpint.cpp and built and the result is:
```bash
659 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ ./bin/exifprint 
Exiv2 version is: 0.27.3
18: 201 199 0 210 0 0 111 209 226 76 64 86 159 80 206 76 64 86 
18: 201 199 0 210 0 0 111 209 226 76 64 86 159 80 206 76 64 86 
```

Running your code without the change in #1360, I had:
```bash
643 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ bin/exifprint 
Exiv2 version is: 0.27.3
17: 201 199 0 210 0 0 111 209 226 76 64 86 159 80 206 76 64 
18: 201 199 0 210 0 0 111 209 226 76 64 86 159 80 206 76 64 86 
644 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```

Please be aware that we have two development branches '0.27-maintenance' and 'master'.  I maintain 0.27-maintenance.  When #1360 is finished and merged, the changes will be promoted to 'master' in time.

**Comment by clanmills** [comment]
@piponazo Two thoughts:

1 Are you going to modify base64encode() as I attempted in #1360 (when we hit the Visual Studio compiler issue)?

2 Will you add @agriggio's example to unitTests/test_futils.cpp?

```
767 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ grep base unitTests/*.cpp
unitTests/test_futils.cpp:TEST(base64encode, encodesValidString)
unitTests/test_futils.cpp:    ASSERT_EQ(1, base64encode(original.c_str(), original.size(), result, encodeLength));
unitTests/test_futils.cpp:TEST(base64encode, doesNotEncodeWithNotBigEnoughResultSize)
unitTests/test_futils.cpp:    ASSERT_EQ(0, base64encode(original.c_str(), original.size(), result, encodeLength));
unitTests/test_futils.cpp:TEST(base64decode, decodesValidString)
unitTests/test_futils.cpp:              base64decode(original.c_str(), result, original.size()));
unitTests/test_slice.cpp:// TODO C++11: test range based for loop
unitTests/test_slice.cpp: * Check that the contents of the slice are ascending via an iterator based for
768 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```

**Comment by piponazo** [comment]
Yes, I will do it, I just need to find the time to do it. Feel free to assing this issue to me :wink:

**Comment by clanmills** [comment]
Wonderful, @piponazo.  This is not urgent as the current #1360 fixes the issue.  My concern was should we close this now, or will you touch it later.  Answer:  You'll touch it later.

**Comment by clanmills** [comment]
Fix submitted. I discovered issues with the base64encode function while working on #1475.  I've chosen to fix this issue in the same PR.
https://github.com/Exiv2/exiv2/pull/1475/commits/87e536dadf22353c0766755205c77b970cb65cfe

**Comment by clanmills** [comment]
Fix is included in #1475.
